src/PerformanceSensitiveAnalyzers/CSharp/Analyzers/CallSiteImplicitAllocationAnalyzer.cs

```diff
@@ -44,61 +42,46 @@ internal sealed class CallSiteImplicitAllocationAnalyzer : AbstractAllocationAna
-                {
-                    CheckNonOverridenMethodOnStruct(methodInfo, reportDiagnostic, invocationExpression);
-                }
+                return;
+            }
 
-                if (!methodInfo.Parameters.IsEmpty && invocationExpression.ArgumentList != null)
-                {
-                    var lastParam = methodInfo.Parameters[^1];
-                    if (lastParam.IsParams)
-                    {
-                        CheckParam(invocationExpression, methodInfo, semanticModel, reportDiagnostic, cancellationToken);
-                    }
-                }
+            var targetMethod = invocationOperation.TargetMethod;
+
+            if (targetMethod.IsOverride)
+            {
+                CheckNonOverridenMethodOnStruct(targetMethod, reportDiagnostic, node);
             }
-        }
 
-        [MethodImpl(MethodImplOptions.NoInlining)]
-        private static void CheckParam(InvocationExpressionSyntax invocationExpression, IMethodSymbol methodInfo, SemanticModel semanticModel, Action<Diagnostic> reportDiagnostic, CancellationToken cancellationToken)
-        {
-            var arguments = invocationExpression.ArgumentList.Arguments;
-            if (arguments.Count == methodInfo.Parameters.Length - 1)
+            bool compilationHasSystemArrayEmpty = !semanticModel.Compilation.GetSpecialType(SpecialType.System_Array).GetMembers("Empty").IsEmpty;
+
+            // Loop on every argument because params argument may not be the last one.
+            //     static void Fun1() => Fun2(args: "", i: 5);
+            //     static void Fun2(int i = 0, params object[] args) {}
+            foreach (var argument in invocationOperation.Arguments)
             {
-                // Up to net45 the System.Array.Empty<T> singleton didn't existed so an empty params array was still causing some memory allocation.
-                if (semanticModel.Compilation.GetSpecialType(SpecialType.System_Array).GetMembers("Empty").IsEmpty)
+                if (argument.ArgumentKind != ArgumentKind.ParamArray)
                 {
-                    reportDiagnostic(invocationExpression.CreateDiagnostic(ParamsParameterRule, EmptyMessageArgs));
+                    continue;
                 }
-                return;
-            }
 
-            if (arguments.Count != methodInfo.Parameters.Length)
-            {
-                reportDiagnostic(invocationExpression.CreateDiagnostic(ParamsParameterRule, EmptyMessageArgs));
-            }
-            else
-            {
-                var lastIndex = arguments.Count - 1;
-                var lastArgumentTypeInfo = semanticModel.GetTypeInfo(arguments[lastIndex].Expression, cancellationToken);
-                if (lastArgumentTypeInfo.Type != null && !lastArgumentTypeInfo.Type.Equals(methodInfo.Parameters[lastIndex].Type))
+                bool isEmpty = (argument.Value as IArrayCreationOperation)?.Initializer.ElementValues.IsEmpty == true;
```

Comment:
We can break out of the loop here if `argument.IsImplicit` is false

Reply:
@genlu I already break at the end of the loop.

Reply:
By breaking here, we can avoid calculating `isEmpty` and (IMO) it makes the code easier to follow. Alternatively, we can rewrite the body of the `foreach` loop to make the flow clearer

```cs
foreach (var argument in invocationOperation.Arguments)
{
    if (argument.ArgumentKind == ArgumentKind.ParamArray && argument.IsImplicit)
    {
        if (!compilationHasSystemArrayEmpty || (argument.Value as IArrayCreationOperation)?.Initializer.ElementValues.IsEmpty == true)
            reportDiagnostic(node.CreateDiagnostic(ParamsParameterRule));

        break;
    }
}
```